### PR TITLE
add DNSSEC validation to the recursive resolver

### LINKED ISSUE #2194: [RFC] DNSSEC validation: configuration syntax
# Motivation

Currently, the recursive resolver (AKA "recursor") component of the `hickory-dns` binary has DNSSEC validation [disabled](https://github.com/hickory-dns/hickory-dns/blob/6334a01430088ead8642cafaee592ec7bf49831f/crates/recursor/src/recursor.rs#L486) and cannot be enabled through its [configuration options](https://github.com/hickory-dns/hickory-dns/blob/6334a01430088ead8642cafaee592ec7bf49831f/crates/server/src/store/recursor/config.rs#L27).

We want to be able to enable DNSSEC validation in the resolver.

## Use cases

We want to support these use cases:

- DNSSEC validation is disabled
- DNSSEC validation is enabled

and when DNSSEC validation is enabled we have these sub- use cases:

- use a *static* user-provided / externally-managed trust anchor
- use a *self-managed* trust anchor, which is automatically updated given an initial trust anchor. see [RFC5011](https://datatracker.ietf.org/doc/html/rfc5011)

# Prior art

## `unbound`

To enable DNSSEC validation and use an externally-managed trust anchor, one has to set the `server.trust-anchor-file` setting to the path to the trust anchor file 

``` text
server:
    trust-anchor-file: /etc/trust-key.key
```

To enable DNSSEC validation and use a self-managed (RFC5011) trust anchor, one has to set the `server.auto-trust-anchor-file` setting to the path to an initial trust anchor file 

``` text
server:
    auto-trust-anchor-file: /etc/trust-key.key
```

An initial trust anchor file can be generated using the `unbound-anchor` tool.

More details about `auto-trust-anchor-file` and `unbound-anchor` can be found in <https://nlnetlabs.nl/documentation/unbound/howto-anchor/>

In both cases, the syntax of the trust anchor file is a newline-separated list of DNS records:

``` text
.	86400	IN	DNSKEY	257 3 7 (.. omitted base64-encoded data ..)
.	86400	IN	DNSKEY	256 3 7 (.. omitted base64-encoded data ..)
```

## BIND (`named`)

`named`'s configuration syntax is more complex. In the `named.conf` file, DNSSEC validation can be enabled / disabled using the `options.dnssec-validation` option:

``` text
options {
     dnssec-validation auto;
}
```

The trust anchor is configured in the file `/etc/bind/bind.keys`, which uses BIND-specific syntax instead of the DNS record syntax.

``` text
trust-anchors {
  . static-key 257 3 7 "(.. omitted base64-encoded data ..)";
  . static-key 256 3 7 "(.. omitted base64-encoded data ..)";
};
```

The `static-key` column indicates whether the key is externally managed (`static-key`) or managed according to RFC5011 (`initial-key`).

# Current state

The recursive resolver component is enabled and configured using a `[[zones]]` section that looks like this:

``` toml
[[zones]]
zone = "."
zone_type = "Hint"
stores = { type = "recursor", roots = "/tmp/root.hints" }
```

Where the `stores` entry corresponds to [`RecursiveConfig`](https://github.com/hickory-dns/hickory-dns/blob/6334a01430088ead8642cafaee592ec7bf49831f/crates/server/src/store/recursor/config.rs#L27) when `type` is `recursor`. 

# Proposal

Although `hickory-dns` configuration file is modeled after (BIND) `named`'s configuration file, I would propose using a configuration syntax closer to `unbound`'s. Namely, adding these two fields to `RecursiveConfig`

``` rust
pub struct RecursiveConfig {
    pub roots: PathBuf,
    pub ns_cache_size: usize,
    pub record_cache_size: usize,
    /// Path to trust anchor file
    pub trust_anchor_file: Option<PathBuf>,
    /// Whether to manage the trust anchor file using RFC5011 or not (default: false)
    pub rfc5011: bool,
}
```

| `trust_anchor_file` | `rfc5011` | DNSSEC validation                                      |
|---------------------|-----------|--------------------------------------------------------|
| `None`              | `_`         | No                                                     |
| `Some(_)`           | `false`   | Yes, `trust_anchor_file` is *static* key               |
| `Some(_)`           | `true`    | Yes, use RFC5011; `trust_anchor_file` is *initial* key |

The syntax of the trust anchor file will be a newline-separated list of DNS records, either DS or DNSKEY records

This proposal can be implemented incrementally. We can add only `trust_anchor_file` first and only support static keys. We can then add `rfc5011` when RFC5011 is implemented (if we want to implement it).

# Alternatives

If we foresee supporting different trust anchor management strategies we could use a single `enum` field instead of an extra boolean field. Something like this:

``` rust
pub struct RecursiveConfig {
    pub roots: PathBuf,
    pub ns_cache_size: usize,
    pub record_cache_size: usize,
    pub dnssec_validation: DnssecValidation,
}

#[non_exhaustive]
pub enum DnssecValidation {
    Disabled,
    // RFC5011
    InitialKey { file: PathBuf },
    // externally-managed key
    StaticKey { file: PathBuf },

    // .. other strategies ..
}
```

### LINKED ISSUE #2234: add DNSSEC validation to `Recursor`
creating this ticket to signal that I'm working on this feature

also see #2194 (though that's the next step: after `Recursor` can do DNSSEC validation, we can expose that via `hickory-server`'s configuration)

### PR DESCRIPTION
**Pull request by japaric**
this is an initial implementation of DNSSEC validation that maximally reuses the existing `DnssecDnsHandle` logic. 

it passes the "happy path" conformance tests but there's more work to be done to pass the tests that exercise the failure modes. namely, fixing the bug reported in #2252 . also, handling the CD (Checking Disabled) bit has not yet been implemented

~~this PR depends on #2245 so I have set that PR's branch as the base for this PR.~~

~~this PR depends on PRs #2257 , #2258 and #2268~~ all those PRs have been merged

~~to avoid this being merged into the wrong branch I'm opening this as a *draft PR* but **it is ready for review**~~ this is now ready to be merged

closes #2194 
closes #2234

### CODE REVIEW

File: crates/proto/src/serialize/txt/parse_rdata.rs [hunk 1]
```diff
@@ -88,6 +88,9 @@ impl RDataParser for RData {
             RecordType::TLSA => Self::TLSA(tlsa::parse(tokens)?),
             RecordType::TXT => Self::TXT(txt::parse(tokens)?),
             RecordType::SIG => return Err(ParseError::from("parsing SIG doesn't make sense")),
+            #[cfg(feature = "dnssec")]
+            RecordType::DNSKEY => Self::DNSSEC(DNSSECRData::DNSKEY(dnskey::parse(tokens)?)),
```
Comment (japaric) [inline crates/proto/src/serialize/txt/parse_rdata.rs:92]: IIUC, hickory-dns is not meant to work with pre-signed zone files like `nsd` does so when this `Parser` API is used to load a zone file into a "file" authority we should continue to reject DNSKEY records.

IOW, I should probably add some flag here that determines whether DNSKEY records should be accepted or not. this `Parser` is also used to parse a "trust anchor" file which is a newline separated list of DNSKEY records; in that case we want the `Parser` to (only) parse DNSKEY records. when we use `Parser` in service of a file authority we want to reject DNSKEY records like the original code was doing.

File: crates/recursor/src/recursor.rs [hunk 2]
```diff
@@ -307,366 +301,106 @@ impl Recursor {
...
+                ))));
+            };
+
+            let this = self.clone();
+            stream::once(async move {
+                // request the DNSSEC records; we'll strip them if not needed on the caller side
+                let do_bit = true;
+                this.resolve(query, Instant::now(), do_bit)
+                    .map_ok(|lookup| {
+                        // XXX `DnssecDnsHandle` will only look at the answer section of the message so
```
Comment (japaric) [inline crates/recursor/src/recursor.rs:392]: this implementation ended up being a bit "forced" as it assumes that this interface will only be used by the `DnssecDnsHandle` newtype, which is the case.

I feel like changing the `DnsHandle` interface would make *this* code reuse of `DnssecDnsHandle` cleaner but `DnsHandle` is used in other places so that would be a much larger change.

Similarly, I considered creating a new "lookup only" interface alternative to `DnsHandle` that would be used as a trait bound in `DnssecDnsHandle` but then I wasn't sure which of the other existing `DnsHandle` implementations should move to new interface and which others should keep using `DnsHandle`.

File: crates/server/Cargo.toml [hunk 3]
```diff
@@ -53,7 +53,7 @@ dnssec = ["hickory-recursor?/dnssec"]
 recursor = ["hickory-recursor"]
 resolver = ["hickory-resolver"]
 sqlite = ["rusqlite"]
-toml = ["dep:basic-toml"]
+toml = ["dep:toml"]
```
Comment (japaric) [inline crates/server/Cargo.toml:56]: as I mentioned in https://github.com/hickory-dns/hickory-dns/issues/2194#issuecomment-2176497799 `basic-toml` cannot deserialize into enum variants that have fields so I switched the toml library to the `toml` crate

File: conformance/packages/dns-test/src/docker/hickory.Dockerfile [hunk 4]
```diff
@@ -1,15 +1,18 @@
...
+        libssl-dev \
         ldnsutils \
         tshark
 
 # `dns-test` will invoke `docker build` from a temporary directory that contains
 # a clone of the hickory repository. `./src` here refers to that clone; not to
 # any directory inside the `dns-test` repository
 COPY ./src /usr/src/hickory
-RUN cargo install --path /usr/src/hickory/bin --features recursor,dnssec-ring --debug && \
+RUN cargo install --path /usr/src/hickory/bin --features recursor,dnssec-openssl --debug && \
```
Comment (japaric) [inline conformance/packages/dns-test/src/docker/hickory.Dockerfile:16]: alternatively, we can continue using the `dnssec-ring` feature and instead switch the algorithm used to sign zone files in conformance tests from RSASHA1_NSEC3 to RSASHA256. any preference?

currently the algorithm is [hardcoded in the `dns-test` library](https://github.com/hickory-dns/hickory-dns/blob/02a8668fd54be5c3b1a9f8d8e79393229ffea2be/conformance/packages/dns-test/src/name_server.rs#L221) but we'll probably want to make that configurable on a test basis. that would let us add a regression test for #2252 if we do switch the algorithm to RSASHA256

Comment (djc) [reply conformance/packages/dns-test/src/docker/hickory.Dockerfile:16]: We hsould probably switch to SHA256 anyway since SHA-1 is deprecated. Would definitely prefer to continue relying on ring (and maybe aws-lc-rs later?) over OpenSSL.

File: crates/recursor/src/recursor.rs [hunk 5]
```diff
@@ -82,7 +59,7 @@ impl RecursorBuilder {
     /// # Panics
     ///
     /// This will panic if the roots are empty.
-    pub fn build(&self, roots: impl Into<NameServerConfigGroup>) -> Result<Recursor, ResolveError> {
+    pub fn build(&self, roots: impl Into<NameServerConfigGroup>) -> Recursor {
```
Comment (djc) [inline crates/recursor/src/recursor.rs]: Nit: would like to extract this change into a separate commit.

File: crates/recursor/src/common.rs [hunk 6]
```diff
@@ -0,0 +1,149 @@
+use std::time::Instant;
```
Comment (djc) [inline crates/recursor/src/common.rs:1]: Not a big fan of files that are this small. Maybe inline the contents of `common` and `dnssec_policy` into `lib.rs`?

Also please keep moving code around in a separate commit.

File: crates/recursor/src/recursor.rs [hunk 7]
```diff
@@ -47,10 +55,10 @@ impl RecursorBuilder {
         self
     }
 
-    /// Enables or disables (DNSSEC) security awareness
+    /// Sets DNSSEC policy
     #[cfg(feature = "dnssec")]
-    pub fn security_aware(&mut self, security_aware: bool) -> &mut Self {
-        self.security_aware = security_aware;
+    pub fn dnssec_policy(&mut self, dnssec_policy: DnssecPolicy) -> &mut Self {
```
Comment (djc) [inline crates/recursor/src/recursor.rs:60]: Would also like a separate commit for the introduction of `DnssecPolicy`.

File: crates/recursor/src/recursor.rs [hunk 8]
```diff
@@ -59,17 +67,12 @@ impl RecursorBuilder {
     /// # Panics
     ///
     /// This will panic if the roots are empty.
-    pub fn build(&self, roots: impl Into<NameServerConfigGroup>) -> Recursor {
-        #[cfg(not(feature = "dnssec"))]
-        let security_aware = false;
-        #[cfg(feature = "dnssec")]
-        let security_aware = self.security_aware;
-
+    pub fn build(&self, roots: impl Into<NameServerConfigGroup>) -> Result<Recursor, Error> {
```
Comment (djc) [inline crates/recursor/src/recursor.rs:70]: If we're going to flip-flop on the fallibility of this method, let's just keep it fallible?

File: crates/recursor/src/recursor.rs [hunk 9]
```diff
@@ -266,8 +301,106 @@ impl Recursor {
...
+                    lookup,
+                    query,
+                ))
+            }
+        }
+    }
+}
+
+enum RecursorEither {
+    NonValidating(RecursorDnsHandle),
```
Comment (djc) [inline crates/recursor/src/recursor.rs:340]: Why doesn't the non-validating variant need a cache?

I dislike `Either` naming because there is no semantic meaning. Can you come up with a better name for this?

File: crates/recursor/src/recursor_dns_handle.rs [hunk 10]
```diff
@@ -138,31 +138,7 @@ impl RecursorDnsHandle {
...
-                        } else {
-                            true
-                        }
-                    });
-
-                let lookup = self.record_cache.insert_records(query, records, now);
-
-                lookup.ok_or_else(|| Error::from("no records found"))
-            }
+            Ok(r) => common::insert_response(r, Some(ns.zone()), &self.record_cache, query, now),
```
Comment (djc) [inline crates/recursor/src/recursor_dns_handle.rs:141]: Please make this a separate commit, too.

File: crates/recursor/src/recursor_dns_handle.rs [hunk 11]
```diff
@@ -27,9 +26,10 @@ use crate::{
 /// Set of nameservers by the zone name
 type NameServerCache<P> = LruCache<Name, RecursorPool<P>>;
 
+#[derive(Clone)]
```
Comment (djc) [inline crates/recursor/src/recursor_dns_handle.rs:29]: Why is this necessary? Are the `RecursorPool` and `DnsLru` cheap to clone? Why not wrap the entire `RecursorDnsHandle` in an `Arc`?

File: crates/proto/src/serialize/txt/parse_rdata.rs (hunk 1 above)
Comment (djc) [reply crates/proto/src/serialize/txt/parse_rdata.rs:92]: In our usage, is there any case where we need to parse both `DNSKEY` records and other kinds of records? If not, we should just split the parser?

File: crates/proto/src/serialize/txt/rdata_parsers/dnskey.rs [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+use crate::serialize::txt::{ParseError, ParseErrorKind, ParseResult};
+
+fn bit_is_set(value: u16, bit: usize) -> bool {
+    let mask = 1 << bit;
+    value & mask == mask
+}
+
+#[allow(deprecated)]
+pub(crate) fn parse<'i, I: Iterator<Item = &'i str>>(mut tokens: I) -> ParseResult<DNSKEY> {
+    let flags_str: &str = tokens
```
Comment (djc) [inline crates/proto/src/serialize/txt/rdata_parsers/dnskey.rs:12]: Nit: let's skip the unnecessary type annotations (in new code)?

File: crates/proto/src/serialize/txt/rdata_parsers/dnskey.rs [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+        .next()
+        .ok_or_else(|| ParseError::from(ParseErrorKind::Message("flags not present")))?;
+    let protocol_str: &str = tokens
+        .next()
+        .ok_or_else(|| ParseError::from(ParseErrorKind::Message("protocol not present")))?;
+    let algorithm_str: &str = tokens
+        .next()
+        .ok_or_else(|| ParseError::from(ParseErrorKind::Message("algorithm not present")))?;
+
+    let flags: u16 = flags_str.parse()?;
```
Comment (djc) [inline crates/proto/src/serialize/txt/rdata_parsers/dnskey.rs:22]: I'd prefer calling this as `u16::from_str(flags_str)` (same for other uses of `.parse()`).

File: crates/server/Cargo.toml (hunk 3 above)
Comment (djc) [reply crates/server/Cargo.toml:56]: IMO that should be a separate commit.

File: crates/server/src/store/recursor/config.rs [hunk 14]
```diff
@@ -78,3 +80,122 @@ fn ns_cache_size_default() -> usize {
...
+
+    /// DNSSEC validation is enabled and will use the chosen `trust_anchor` set of keys
+    #[cfg(feature = "dnssec")]
+    ValidateWithStaticKey {
+        /// set to `None` to use built-in trust anchor
+        path: Option<PathBuf>,
+    },
+}
+
+impl Default for DnssecPolicyConfig {
```
Comment (djc) [inline crates/server/src/store/recursor/config.rs:101]: Can this be derived?

File: crates/server/src/store/recursor/config.rs [hunk 15]
```diff
@@ -78,3 +80,122 @@ fn ns_cache_size_default() -> usize {
...
+    fn default() -> Self {
+        Self::SecurityUnaware
+    }
+}
+
+impl DnssecPolicyConfig {
+    pub(crate) fn load(&self) -> Result<DnssecPolicy, String> {
+        let policy = match self {
```
Comment (djc) [inline crates/server/src/store/recursor/config.rs:109]: Nit: inline the `policy` binding?

File: crates/server/src/store/recursor/config.rs [hunk 16]
```diff
@@ -78,3 +80,122 @@ fn ns_cache_size_default() -> usize {
...
+    use crate::proto::rr::dnssec::rdata::DNSKEY;
+    use crate::proto::rr::dnssec::PublicKeyEnum;
+    use crate::proto::rr::{RecordData, RecordType};
+
+    let parser = Parser::new(input, None, Some(Name::root()));
+    let (_, records) = parser.parse().map_err(|e| e.to_string())?;
+
+    let mut trust_anchor = TrustAnchor::new();
+    for (key, rrset) in records {
+        if key.record_type == RecordType::DNSKEY {
```
Comment (djc) [inline crates/server/src/store/recursor/config.rs:146]: Nit: invert and continue early?

File: crates/server/src/store/recursor/authority.rs [hunk 17]
```diff
@@ -172,4 +172,12 @@ impl LookupObject for RecursiveLookup {
     fn take_additionals(&mut self) -> Option<Box<dyn LookupObject>> {
         None
     }
+
+    fn dnssec_validated(&self) -> bool {
+        // XXX what should be the behavior if the lookup yields no records?
```
Comment (djc) [inline crates/server/src/store/recursor/authority.rs:177]: Does this need discussion?

File: crates/recursor/src/recursor_dns_handle.rs (hunk 11 above)
Comment (japaric) [reply crates/recursor/src/recursor_dns_handle.rs:29]: > Why is this necessary?

`DnsHandle` is a subtrait of `Clone` and we need to implement `DnsHandle` to put this type inside `DnssecDnsHandle<_>`

> Are the RecursorPool and DnsLru cheap to clone?

`RecursorPool` has several `Arc` fields and  `DnsLru` is esentially an `Arc<Mutex<_>>` so, yes

> Why not wrap the entire RecursorDnsHandle in an Arc?

doable but that would result in a nested `Arc` which ... I guess it's fine? I don't think either way has much impact on performance, either CPU or memory usage.

File: crates/proto/src/serialize/txt/parse_rdata.rs (hunk 1 above)
Comment (japaric) [reply crates/proto/src/serialize/txt/parse_rdata.rs:92]: Well, `unbound` supports trust anchor files that contain both `DNSKEY` and `DS` records. Meaning that `unbound` can complete DNSSEC validation with just a DS record in the trust anchor file.

In their current state, `DnssecDnsHandle` and `TrustAnchor` only support DNSSEC validation using public key material (`DNSKEY` information) so a trust anchor file with just DS records would not work and should be rejected.

So, yeah. It's probably best to have separate parsers: the general `Parser` for regular zone files and a `TrustAnchorParser` that only accepts newline separated DNSKEY records (and maybe also DS records in the future).

File: crates/recursor/src/recursor_dns_handle.rs (hunk 11 above)
Comment (djc) [reply crates/recursor/src/recursor_dns_handle.rs:29]: I guess it's fine, no need to change it on my account.

File: crates/server/src/store/recursor/authority.rs (hunk 17 above)
Comment (japaric) [reply crates/server/src/store/recursor/authority.rs:177]: I think it needs a better understanding of the RFCs and an accompanying conformance test.

I think the condition mentioned here would happen if a query yields an empty answer section (NXDOMAIN), DNSSEC validation is enabled but both NSEC and NSEC3 are disabled. "does the AD flag get set in the response or not?" is what needs to be researched, either in RFCs or by testing what other implementations do.

I have that in my TODO list. I can also open issues for missing / pending conformance tests if you think that's worthwhile.

Comment (djc) [reply crates/server/src/store/recursor/authority.rs:177]: If it's on your list I'm fine merging it as is. Just wanted to mark it with a discussion to get a bit more context on how/when this will be followed up on.

File: crates/proto/src/serialize/txt/parse_rdata.rs (hunk 1 above)
Comment (japaric) [reply crates/proto/src/serialize/txt/parse_rdata.rs:92]: I have submitted this whole separate parser idea in #2257

**Comment by japaric** [comment]
Addressed all review comments, I believe. Note that this version includes two other open PRs as cherry picks. I'll remove them once those PRs have been merged

File: crates/recursor/src/recursor_dns_handle.rs [hunk 18]
```diff
@@ -0,0 +1,295 @@
+use std::net::SocketAddr;
```
Comment (djc) [inline crates/recursor/src/recursor_dns_handle.rs]: This doesn't look like it belongs in this commit.

Comment (japaric) [reply crates/recursor/src/recursor_dns_handle.rs]: good catch! fixed

File: crates/recursor/src/lib.rs [hunk 19]
```diff
@@ -54,6 +56,15 @@ pub enum DnssecPolicy {
     /// DNSSEC validation is disabled; DNSSEC records will be requested and processed
     #[cfg(feature = "dnssec")]
     ValidationDisabled,
+
+    /// DNSSEC validation is enabled and will use the chosen `trust_anchor` set of keys
+    #[cfg(feature = "dnssec")]
+    ValidateWithStaticKey {
+        /// set to `None` to use built-in trust anchor
+        trust_anchor: Option<TrustAnchor>,
```
[the remark this answers is not in the document]
Comment (japaric) [reply crates/recursor/src/lib.rs]: I like that idea.

It does require changes elsewhere because this `trust_anchor` is used to initialize a `DnssecDnsHandle` and its constructor expects a `TrustAnchor` by value ... which then immediately wraps in an `Arc` for storage in one of its fields.

I submitted the `Arc` changes in a separate commit

**Comment by japaric** [comment]
rebased to fix merge conflict

note that I have cherry picked PR #2268 because `dns_test::resolver::terminate_hickory_works` started failing after the configuration file changed ("expose DnssecPolicy in RecursorConfig")

until the 3 cherry picked PRs have been merged, I'll keep this PR in the draft state but it is ready for review

**Comment by japaric** [comment]
amended a commit ("expose DnssecPolicy in RecursorConfig") to fix a clippy warning

**Comment by yonas** [comment]
@japaric I just tested the recurser. It doesn't pickup the `--roots` CLI flag.

**Comment by japaric** [comment]
@yonas thanks for testing. that sounds like it may be a pre-existing bug.

the end-to-end tests here all configure the root hints in `named.toml`; see the `conformance/packages/dns-test/src/templates/hickory.resolver.toml.jinja` file in this branch / PR. 

did you set up a `name.toml` configuration file and additionally passed the `--roots` flag to `hickory-dns`? I'm not sure what the precedence rules are in that particular case since I didn't touch CLI code in this PR. if that sort of usage does not produce a warning / error then could you open an issue about that? because to me that sounds like it should.

relatedly, we could use some docs on setting up `hickory-dns` as a recursive resolver. I floated the idea in https://github.com/hickory-dns/hickory-dns.github.io/issues/2 but I'm sure how well publicized the book currently is. maybe those docs can go in the README for now, if that has more visibility. in any case, it would be good to fix #2252 before the DNSSEC feature gets more publicity, IMO.

wait, `hickory-dns` does not have a `--roots` flags. are you using `util/src/bin/recurse.rs`?  that tool has not been touched in this PR. I don't know what's the plan with `hickory-util` so DNSSEC has not been exposed there.

**Comment by yonas** [comment]
No, I did not setup `name.toml`

Yes, I used the `recurse` binary.

Which binary does this PR affect?

**Comment by japaric** [comment]
> Which binary does this PR affect?

`hickory-dns` (DNS server) which resides in `bin/src/hickory-dns.rs` and you can build with `cargo build -p hickory-dns`. the artifact will then be in `target/debug/hickory-dns`. you probably want to run that with `-p 1053 -c my-named.toml` because the defaults require root permissions (to bind port 53 and to create a file in `/etc/named.toml`). if you need help setting that up, I would suggest asking in the [discord server](https://discord.gg/89nxE4n) to keep this thread on topic.

**Comment by bluejekyll** [comment]
@japaric is this ready to merge in your opinion? I know there was a lot of commentary back and forth.

**Review by djc — APPROVED** [review]

File: conformance/packages/dns-test/src/templates/hickory.resolver.toml.jinja [hunk 20]
```diff
@@ -1,5 +1,4 @@
 [[zones]]
 zone = "."
 zone_type = "Hint"
-stores = { type = "recursor", roots = "/etc/root.hints", security_aware = true }
-enable_dnssec = {{ use_dnssec }}
```
Comment (justahero) [inline conformance/packages/dns-test/src/templates/hickory.resolver.toml.jinja:5]: If my understanding is correct the `enable_dnssec` is not used by the `Resolver`, the DNSSEC logic is handledby the given `DnssecPolicy` here? Does this in any way affect the example`toml` files in `tests/test-data/test_configs` &  should they be updated as well?

Comment (japaric) [reply conformance/packages/dns-test/src/templates/hickory.resolver.toml.jinja:5]: `enable_dnssec` is only used by the `file` and `sqlite` type stores (which must be used with the `Primary` and `Secondary` `zone_type`s). it is not used by the `recursor` (`Hint`) nor the `forward`-type stores. 

IOW, this `enable_dnssec` had no effect in this file. I added a while ago when I did not understand the structure of the `named.toml` file -- which then prompted me to suggest RFC #2195

So, you should not understand `enable_dnssec = true` must now be replaced with `dnssec_policy = Something` but rather as the `enable_dnssec`, which was a no-op, is getting removed (clean-up); *and* the use of `security_aware` is being replaced with `dnssec_policy`.

Therefore, the `toml` files under `test_configs` do not need to be updated as part of this PR.

**Review by justahero — APPROVED** [review]

**Comment by japaric** [comment]
rebased to fix a merge conflict caused by a docstring change

@bluejekyll yes, this is ready to be merged :rocket:

**Comment by djc** [comment]
This failed CI on main:

https://github.com/hickory-dns/hickory-dns/actions/runs/9777369448/job/26991905906#step:5:2056

**Comment by japaric** [comment]
that's a known spurious failure that happens very infrequently: https://github.com/ferrous-systems/dnssec-tests/issues/55

I'll port that issue over here
